src/Lecture2.hs

```diff
@@ -52,7 +68,9 @@ zero, you can stop calculating product and return 0 immediately.
 84
 -}
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
+lazyProduct = foldr step (1 :: Int)
+  where step :: Int -> Int -> Int
+        step x y = if x == 0 then 0 else x*y
```

Comment:
This is a neat implementation! I see that it indeed does stop correctly on zero 🙂 

It contains a space leak still but I cover this in Lecture 4 only